### PR TITLE
Run the secrets-provider with a limited user instead of with root

### PR DESCRIPTION
**Pull request by sigalsax**
- A container running as root is not a good security practice so this should be avoided
- Add group permissions in Conjur directories to avoid UID mismatch upon deploying to Openshift

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -37,9 +37,29 @@ COPY --from=busybox /bin/mkdir /bin/mkdir
...
+    chmod 770 /etc/conjur/ssl \
+              /run/conjur
 
-COPY --from=secrets-provider-builder /opt/secrets-provider-for-k8s/secrets-provider /bin
+USER secrets-provider
 
-CMD ["secrets-provider"]
+COPY --from=secrets-provider-builder /opt/secrets-provider-for-k8s/secrets-provider /usr/local/bin/
+
+ENTRYPOINT [ "/usr/local/bin/secrets-provider"]
```
Comment (orenbm) [inline Dockerfile:65]: missing new line

File: Dockerfile [hunk 2]
```diff
@@ -3,7 +3,7 @@ MAINTAINER CyberArk Software, Ltd.
 
 ENV GOOS=linux \
     GOARCH=amd64 \
-    CGO_ENABLED=0
+    CGO_ENABLED=1
```
Comment (orenbm) [inline Dockerfile:6]: why do we need `cgo_enabled`? are you sure it is required?

File: Dockerfile (hunk 1 above)
Comment (orenbm) [inline Dockerfile:65]: why did you change from `CMD` to `ENTRYPOINT`? can you challenge this?

File: Dockerfile [hunk 3]
```diff
@@ -23,7 +23,7 @@ RUN go build -a -installsuffix cgo -o secrets-provider ./cmd/secrets-provider
 FROM busybox
 
 # =================== MAIN CONTAINER ===================
-FROM scratch as secrets-provider
+FROM alpine:3.8 as secrets-provider
```
Comment (orenbm) [inline Dockerfile:26]: why do we use 3.8? why not `latest`? why not a newer version?

i'm not saying it's incorrect, just want to verify you put thought into it and chose the best one.

**Comment by orenbm** [comment]
i actually prefer the title `Run the secrets-provider with a limited user instead with root` as it better says what the change was

File: Dockerfile (hunk 1 above)
Comment (sigalsax) [reply Dockerfile:65]: CMD sets default command and can be overwritten from command line when docker container runs (with `docker run -it <image> bash`). ENTRYPOINT command and parameters will not be overwritten from command line. So I guess in our case it doesn't really matter but thought I would go with a command that cannot be overwritten

File: Dockerfile (hunk 3 above)
Comment (sigalsax) [reply Dockerfile:26]: good point. i did it for uniformity across projects but 3.11 is available and the local build passes so I will use that one

File: Dockerfile (hunk 2 above)
Comment (sigalsax) [reply Dockerfile:6]: I am not sure about this one. Other projects used it compile source code in Go and C

**Comment by orenbm** [comment]
@sigalsax please add a changelog entry for this change.

File: Dockerfile (hunk 2 above)
Comment (sigalsax) [reply Dockerfile:6]: do you think we need this offering here?

Comment (orenbm) [reply Dockerfile:6]: please don't enable cgo if we don't have to. if the build passes without it then leave it like that.

File: Dockerfile [hunk 4]
```diff
@@ -1,9 +1,8 @@
 FROM golang:1.13 as secrets-provider-builder
-MAINTAINER CyberArk Software, Ltd.
+MAINTAINER CyberArk Software Ltd.
 
 ENV GOOS=linux \
-    GOARCH=amd64 \
-    CGO_ENABLED=0
```
Comment (orenbm) [inline Dockerfile:6]: why was this removed? it should still be here (although it's the default value, it's best to have it for explicitness)

File: Dockerfile [hunk 5]
```diff
@@ -23,8 +22,8 @@ RUN go build -a -installsuffix cgo -o secrets-provider ./cmd/secrets-provider
 FROM busybox
 
 # =================== MAIN CONTAINER ===================
-FROM scratch as secrets-provider
-MAINTAINER CyberArk Software, Ltd.
+FROM alpine:3.11 as secrets-provider
```
Comment (orenbm) [inline Dockerfile:25]: why not use `latest`?

File: Dockerfile [hunk 6]
```diff
@@ -37,9 +36,30 @@ COPY --from=busybox /bin/mkdir /bin/mkdir
...
+    # match our UID when we try to write files to them
+    chmod 770 /etc/conjur/ssl \
+              /run/conjur
 
-COPY --from=secrets-provider-builder /opt/secrets-provider-for-k8s/secrets-provider /bin
+USER secrets-provider
+
+COPY --from=secrets-provider-builder /opt/secrets-provider-for-k8s/secrets-provider /usr/local/bin/
+
+CMD [ "/usr/local/bin/secrets-provider"]
```
Comment (orenbm) [inline Dockerfile:64]: so `CMD` is better than `ENTRYPOINT`? it seemed that the latter is safer from your previous explanation.

File: Dockerfile (hunk 5 above)
Comment (sigalsax) [reply Dockerfile:25]: because I don't like the idea of automatically pulling latest for the same reason you thought; risky. But open to `latest` if you think this is the way to go

Comment (orenbm) [reply Dockerfile:25]: i think that our nightly automation is good enough but i can live with 3.11.

File: Dockerfile (hunk 1 above)
Comment (sigalsax) [reply Dockerfile]: I didn't want secrets-provider to be overridden but CMD is safer...

File: Dockerfile (hunk 6 above)
Comment (sigalsax) [reply Dockerfile]: safer is CMD I would say but I was thinking that we didn't want `secrets-provider` to be overriden because this is the purpose of the build, to run the application. Changing it back because CMD _is_ safer.

Comment (sigalsax) [reply Dockerfile]: ENTRYPOINT cannot be overriden and CMD can. What do you means in terms of "Safer"?
